src/Entities/Player.cpp

```diff
@@ -474,61 +451,87 @@ void cPlayer::SetTouchGround(bool a_bTouchGround)
 		return;
 	}
 
-	m_bTouchGround = a_bTouchGround;
-
-	if (!m_bTouchGround)
-	{
-		if (GetPosY() > m_LastJumpHeight)
-		{
-			m_LastJumpHeight = static_cast<float>(GetPosY());
-		}
-		cWorld * World = GetWorld();
-		if ((GetPosY() >= 0) && (GetPosY() < cChunkDef::Height))
-		{
-			BLOCKTYPE BlockType = World->GetBlock(POSX_TOINT, POSY_TOINT, POSZ_TOINT);
-			if (BlockType != E_BLOCK_AIR)
-			{
-				m_bTouchGround = true;
-			}
-			if (
-				(BlockType == E_BLOCK_WATER) ||
-				(BlockType == E_BLOCK_STATIONARY_WATER) ||
-				(BlockType == E_BLOCK_LADDER) ||
-				(BlockType == E_BLOCK_VINES)
+	/* Not pretty looking, and is more suited to wherever server-sided collision detection is implemented.
+	The following condition sets on-ground-ness if
+		The player isn't swimming or flying (client hardcoded conditions) and
+		they're on a block (Y is exact) - ensure any they could be standing on (including on the edges) is solid or
+		they're on a slab (Y significand is 0.5) - ditto with slab check
+		they're on a snow layer (Y divisible by 0.125) - ditto with snow layer check
+	*/
+
+	static const auto HalfWidth = GetWidth() / 2;
+	if (
+		!IsSwimming() && !IsFlying() &&
+		(
+			(
+				((GetPosY() >= 1) && ((GetPosY() - POSY_TOINT) <= std::numeric_limits<double>::epsilon())) &&
+				(
+					cBlockInfo::IsSolid(GetWorld()->GetBlock((GetPosition() + Vector3d(0, -1, 0)).Floor())) ||
+					cBlockInfo::IsSolid(GetWorld()->GetBlock((GetPosition() + Vector3d(HalfWidth, -1, 0)).Floor())) ||
+					cBlockInfo::IsSolid(GetWorld()->GetBlock((GetPosition() + Vector3d(-HalfWidth, -1, 0)).Floor())) ||
+					cBlockInfo::IsSolid(GetWorld()->GetBlock((GetPosition() + Vector3d(0, -1, HalfWidth)).Floor())) ||
+					cBlockInfo::IsSolid(GetWorld()->GetBlock((GetPosition() + Vector3d(0, -1, -HalfWidth)).Floor()))
+				)
+			) ||
+			(
+				((GetPosY() >= POSY_TOINT) && ((GetPosY() - (POSY_TOINT + 0.5)) <= std::numeric_limits<double>::epsilon())) &&
+				(
+					cBlockSlabHandler::IsAnySlabType(GetWorld()->GetBlock((GetPosition() + Vector3d(0, 0, 0)).Floor())) ||
+					cBlockSlabHandler::IsAnySlabType(GetWorld()->GetBlock((GetPosition() + Vector3d(HalfWidth, 0, 0)).Floor())) ||
+					cBlockSlabHandler::IsAnySlabType(GetWorld()->GetBlock((GetPosition() + Vector3d(-HalfWidth, 0, 0)).Floor())) ||
+					cBlockSlabHandler::IsAnySlabType(GetWorld()->GetBlock((GetPosition() + Vector3d(0, 0, HalfWidth)).Floor())) ||
+					cBlockSlabHandler::IsAnySlabType(GetWorld()->GetBlock((GetPosition() + Vector3d(0, 0, -HalfWidth)).Floor()))
+				)
+			) ||
+			(
+				(fmod(GetPosY(), 0.125) <= std::numeric_limits<double>::epsilon()) &&
```

Comment:
That's why I said "named" - name it whatever, `EPS`, I don't care, but make the value reasonable, it's definitely not a `std::numeric_limits::epsilon`